### PR TITLE
Add support for single-document collections

### LINKED ISSUE #584: Single document collections
If you're building out a relatively 'static' site, it would be handy to specify that a collection only has a single document, so in Publish it skips the `list` and goes directly to the `edit` view.

A good example would be pages that have static elements made up of different fields that don't change that often e.g., homepage consisting of `headline`, `introduction` etc

### PR DESCRIPTION
**Pull request by ohmoses**
Closes #584 

If `settings.publish.isSingleDocument` is set to `true` in the collection schema, Publish will show the document editor (instead of the document list) on route `/collectionName` (or `/groupName/collectionName`). If there are multiple documents in the collection, Publish will operate on the first one.

I made a small change to enable `connectRedux` to accept a `mapState` function, so that in `ReferenceSelectView`, `DocumentListView`, and `DocumentEditView`, I could move the logic for creating `contentKey`s, `documentId`, etc., to `mapState` and pass them to the components.

### CODE REVIEW

File: app/views/DocumentEditView/DocumentEditView.jsx [hunk 1]
```diff
@@ -28,28 +28,27 @@ class DocumentEditView extends React.Component {
...
-    const {documentId} = route.params
-    const document = state.document[this.contentKey] || {}
-    const oldDocument = oldState.document[this.contentKey] || {}
+    const document = state.document[contentKey] || {}
+    const oldDocument = oldState.document[contentKey] || {}
     const isSaving = (oldDocument.saveAttempts || 0) < document.saveAttempts
 
     // Are there unsaved changes?
     if (
+      !document.isLoading &&
```
Comment (ohmoses) [inline app/views/DocumentEditView/DocumentEditView.jsx]: This fixes a minor bug: if there previously were unsaved changes, but now there are not, then in the short amount of time before the document is loaded (with `wasLoadedFromLocalStorage` being false), the previous version of the document (with `wasLoadedFromLocalStorage` true) is grabbed from Redux, which causes the "unsaved changes" notification to fire.

**Comment by ohmoses** [comment]
A point for discussion: I removed the "Delete" button, because it didn't make much sense here, but maybe there should be something like "clear all fields" button?

File: app/views/ReferenceSelectView/ReferenceSelectView.jsx [hunk 2]
```diff
@@ -195,25 +169,34 @@ class ReferenceSelectView extends React.Component {
...
     const pageNumber = parsedPage.toString() === page ? parsedPage : undefined
 
-    // Getting the schema for the parent collection.
-    const collection =
-      collectionName === Constants.MEDIA_COLLECTION_SCHEMA.slug
-        ? Constants.MEDIA_COLLECTION_SCHEMA
-        : api.collections.find(collection => {
-            return collection.slug === route.params.collection
-          })
+    if (!(isNewDocument || isSingleDocument || documentId)) {
```
Comment (eduardoboucas) [inline app/views/ReferenceSelectView/ReferenceSelectView.jsx]:
```suggestion
    if (!isNewDocument && !isSingleDocument && !documentId)) {
```

... is easier on the brain, I think?

File: app/views/ReferenceSelectView/ReferenceSelectView.jsx [hunk 3]
```diff
@@ -384,6 +366,21 @@ class ReferenceSelectView extends React.Component {
         }}
       />
     )
+
+    if (isSingleDocument) {
+      // In this case the documentId is not in the url, so we have to fetch the
```
Comment (eduardoboucas) [inline app/views/ReferenceSelectView/ReferenceSelectView.jsx]:
```suggestion
      // In this case the documentId is not in the URL, so we have to fetch the
```

🤓

File: app/views/ReferenceSelectView/ReferenceSelectView.jsx [hunk 4]
```diff
@@ -449,6 +446,54 @@ class ReferenceSelectView extends React.Component {
...
+  } = ownProps
+
+  const collection =
+    params.collection === Constants.MEDIA_COLLECTION_SCHEMA.slug
+      ? Constants.MEDIA_COLLECTION_SCHEMA
+      : state.app.config.api.collections.find(collection => {
+          return collection.slug === params.collection
+        })
+
+  const isNewDocument = /\/new\/?/.test(path) && !params.documentId
```
Comment (eduardoboucas) [inline app/views/ReferenceSelectView/ReferenceSelectView.jsx]: Is this a good check? It correctly matches `/articles/new`, but also matches `/newyork/12345`.

File: app/views/DocumentListView/DocumentListView.jsx [hunk 5]
```diff
@@ -229,8 +214,36 @@ class DocumentListView extends React.Component {
...
+              <Main>
+                <SpinningWheel />
+              </Main>
+            </Page>
+          )}
+          onNetworkError={this.handleNetworkError.bind(this)}
+          onRender={({documents}) => (
+            <DocumentEditView
+              {...this.props}
+              isSingleDocument
```
Comment (eduardoboucas) [inline app/views/DocumentListView/DocumentListView.jsx]: Sort props alphabetically, please. 🤓

File: app/containers/App/App.jsx [hunk 6]
```diff
@@ -280,95 +331,17 @@ class App extends React.Component {
...
-          <Route
-            isSignedIn={isSignedIn}
-            exact
-            component={ReferenceSelectView}
-            config={state.app.config}
-            mustBeSignedIn
-            path={`/:group${REGEX_SLUG}/:collection${REGEX_SLUG}/:documentId${REGEX_DOCUMENT_ID}/select/:referenceField/:page${REGEX_NUMBER}?`}
-          />
+          {collectionRoutes.map(({path, component}) => (
+            <Route
```
Comment (eduardoboucas) [inline app/containers/App/App.jsx]: Do you mind sorting the props alphabetically here, please? (I know they weren't before.)

File: app/views/ReferenceSelectView/ReferenceSelectView.jsx [hunk 7]
```diff
@@ -449,6 +446,54 @@ class ReferenceSelectView extends React.Component {
...
+  } = ownProps
+
+  const collection =
+    params.collection === Constants.MEDIA_COLLECTION_SCHEMA.slug
+      ? Constants.MEDIA_COLLECTION_SCHEMA
+      : state.app.config.api.collections.find(collection => {
+          return collection.slug === params.collection
+        })
+
+  const isNewDocument = /\/new(\/|$)/.test(path) && !params.documentId
```
Comment (eduardoboucas) [inline app/views/ReferenceSelectView/ReferenceSelectView.jsx]: @ohmoses is there ever a case where checking for `params.documentId` is not enough to determine whether we're creating a new document?

**Pull request merged**